File: src-typescript/ch06/RemoteLoaderWithMacro.ts
import {RemoteControlWithUndo} from "./RemoteControlWithUndo";
import {Light} from "./Light";
import {GarageDoor} from "./GarageDoor";
import {Stereo} from "./Stereo";
import {LightOnCommand} from "./LightOnCommand";
import {LightOffCommand} from "./LightOffCommand";
import {GarageDoorUpCommand} from "./GarageDoorUpCommand";
import {GarageDoorDownCommand} from "./GarageDoorDownCommand";
import {StereoOnWithCdCommand} from "./StereoOnWithCdCommand";
import {StereoOffCommand} from "./StereoOffCommand";
import {Command} from "./interface/Command";
import {MacroCommand} from "./MacroCommand";

class RemoteLoaderWithMacro {
    static main(): void {
        const remoteControl: RemoteControlWithUndo = new RemoteControlWithUndo();
        const light: Light = new Light("Living Room");
        const garageDoor: GarageDoor = new GarageDoor("Living Room");
        const stereo: Stereo = new Stereo("Living Room");

        const lightOn: LightOnCommand = new LightOnCommand(light);
        const lightOff: LightOffCommand = new LightOffCommand(light);
        const garageDoorUp: GarageDoorUpCommand = new GarageDoorUpCommand(garageDoor);
        const garageDoorDown: GarageDoorDownCommand = new GarageDoorDownCommand(garageDoor);
        const stereoOnWithCd: StereoOnWithCdCommand = new StereoOnWithCdCommand(stereo);
        const stereoOff: StereoOffCommand = new StereoOffCommand(stereo);

        const partyOn: Command[] = [lightOn, stereoOnWithCd, garageDoorUp];
        const partyOff: Command[] = [lightOff, stereoOff, garageDoorDown];

        const partyOnMacro: MacroCommand = new MacroCommand(partyOn);
        const partyOffMacro: MacroCommand = new MacroCommand(partyOff);

        remoteControl.setCommands(0, partyOnMacro, partyOffMacro);

        remoteControl.className();
        console.log(" --- 메크로 ON --- ");
        remoteControl.onButtonWasPushed(0);
        console.log(" --- 메크로 OFF --- ");
        remoteControl.offButtonWasPushed(0);
    }
}

RemoteLoaderWithMacro.main();

File: src/ch06/RemoteLoaderWithMacro.java
package ch06;

public class RemoteLoaderWithMacro {

    public static void main(String[] args) {
        RemoteControlWithUndo remoteControl = new RemoteControlWithUndo();
        Light light = new Light("Living Room");
        GarageDoor garageDoor = new GarageDoor("Living Room");
        Stereo stereo = new Stereo("Living Room");

        LightOnCommand lightOn = new LightOnCommand(light);
        GarageDoorUpCommand garageDoorUp = new GarageDoorUpCommand(garageDoor);
        StereoOnWithCdCommand stereoOn = new StereoOnWithCdCommand(stereo);

        LightOffCommand lightOff = new LightOffCommand(light);
        GarageDoorDownCommand garageDoorDown = new GarageDoorDownCommand(garageDoor);
        StereoOffCommand stereoOff = new StereoOffCommand(stereo);

        Command[] partyOn = {lightOn, stereoOn, garageDoorUp};
        Command[] partyOff = {lightOff, stereoOff, garageDoorDown};

        MacroCommand partyOnMacro = new MacroCommand(partyOn);
        MacroCommand partyOffMacro = new MacroCommand(partyOff);

        remoteControl.setCommands(0, partyOnMacro, partyOffMacro);

        System.out.println(remoteControl);
        System.out.println("--- 메크로 ON ---");
        remoteControl.onButtonWasPushed(0);
        System.out.println("--- 메크로 OFF ---");
        remoteControl.offButtonWasPushed(0);
    }
}
